test/mainnet-fork/SpkFarm.t.sol

```diff
@@ -0,0 +1,172 @@
+    function test_depositUSDSToSPKFarm() external {
+        bytes32 depositKey = RateLimitHelpers.makeAssetKey(
+            mainnetController.LIMIT_SPK_FARM_DEPOSIT(),
+            spkFarm
+        );
+
+        deal(address(usds), address(almProxy), 1_000_000e18);
+
+        assertEq(rateLimits.getCurrentRateLimit(depositKey), 10_000_000e18);
+
+        assertEq(usds.balanceOf(address(almProxy)),                  1_000_000e18);
+        assertEq(ISPKFarmLike(spkFarm).balanceOf(address(almProxy)), 0);
+
+        vm.prank(relayer);
+        mainnetController.depositUSDSToSPKFarm(spkFarm, 1_000_000e18);
+
+        assertEq(rateLimits.getCurrentRateLimit(depositKey), 9_000_000e18);
+
+        assertEq(usds.balanceOf(address(almProxy)),                  0);
+        assertEq(ISPKFarmLike(spkFarm).balanceOf(address(almProxy)), 1_000_000e18);
+    }
+
+}
+
+contract MainnetControllerSPKFarmWithdrawFailureTests is MainnetControllerSPKFarmTestBase {
+
+    function test_withdrawUSDSFromSPKFarm_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        mainnetController.withdrawUSDSFromSPKFarm(spkFarm, 1_000_000e18);
+    }
+
+    function test_withdrawUSDSFromSPKFarm_zeroMaxAmount() external {
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/zero-maxAmount");
+        mainnetController.withdrawUSDSFromSPKFarm(makeAddr("fake-spk-farm"), 0);
+    }
+
+    function test_withdrawUSDSFromSPKFarm_rateLimitsBoundary() external {
+        bytes32 key = RateLimitHelpers.makeAssetKey(
+            mainnetController.LIMIT_SPK_FARM_WITHDRAW(),
+            spkFarm
+        );
+
+        vm.prank(Ethereum.SPARK_PROXY);
+        rateLimits.setRateLimitData(key, 1_000_000e18, uint256(1_000_000e18) / 1 days);
+
+        deal(address(usds), address(almProxy), 1_000_000e18);
+        vm.startPrank(relayer);
+        mainnetController.depositUSDSToSPKFarm(spkFarm, 1_000_000e18);
+
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        mainnetController.withdrawUSDSFromSPKFarm(spkFarm, 1_000_000e18 + 1);
+
+        mainnetController.withdrawUSDSFromSPKFarm(spkFarm, 1_000_000e18);
+    }
+
+}
+
+contract MainnetControllerSPKFarmWithdrawSuccessTests is MainnetControllerSPKFarmTestBase {
+
+    function test_withdrawUSDSFromSPKFarm() external {
+        bytes32 withdrawKey = RateLimitHelpers.makeAssetKey(
+            mainnetController.LIMIT_SPK_FARM_WITHDRAW(),
+            spkFarm
+        );
+
+        deal(address(usds), address(almProxy), 1_000_000e18);
+        vm.prank(relayer);
+        mainnetController.depositUSDSToSPKFarm(spkFarm, 1_000_000e18);
+
+        assertEq(rateLimits.getCurrentRateLimit(withdrawKey), 10_000_000e18);
+
+        assertEq(usds.balanceOf(address(almProxy)),                  0);
+        assertEq(ISPKFarmLike(spkFarm).balanceOf(address(almProxy)), 1_000_000e18);
+
```

Comment:
Warp here and then assert that an SPK balance was claimed

Reply:
done